### ISSUE TITLE
A converged NUTS fit is SIGKILLed in post-processing; deterministic by SSP grid size

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#2356]
A completed NUTS fit is destroyed after it succeeds. The sampler finishes with good
diagnostics (frequently **0 divergences in 2400 draws**), and the process is then
SIGKILLed (exit 137) during post-processing, before anything reaches disk. The work is
lost silently: no traceback, no Python-level error, the log simply stops.

It is deterministic by SSP grid size, so it looks healthy on small grids and fails
every time on large ones.

## Impact

A 120-cell production grid produced **8 usable cells in four hours**. Twenty-one fits
converged and were discarded, all on the large-grid configurations.

| configuration | SSP grid | file | outcome |
|---|---|---|---|
| PARSEC/MILES wNE | `ssp_prsc_miles_chabrier_wNE...` | 64 MB | **5/5 succeeded** |
| BPASS stars-only | `bpss_stars_c3k_a_chabrier` | 3.1 MB | 3/4 succeeded |
| BC03/STELIB | `bc03_pdva_stelib_chabrier` | 35 MB | **0/5** |
| FSPS MIST/C3K | `fsps_mist_c3k_a_chabrier` | 109 MB | **0/5** |
| ProGeny MIST/C3K | `pgny_mist_c3k_chabrier` | 104 MB | **0/5** |

That 0/5 vs 5/5 split across 29 attempts is the most robust signal in this report.

## Reproduction

48 GB machine, macOS, CPU (`JAX_PLATFORMS=cpu`), one fit at a time.

```
python -m analysis.paper1.fit_one --galaxy 9884 --config IV \
    --method mcmc_nuts_fast --out <dir>
```

Configuration IV is a 4-free-parameter constant-SFH model on `fsps_mist_c3k_a_chabrier`.
The fit converges (0/2400 divergences) and the process never writes its output.

## Measurements

RSS sampled every 2 s, plus `resource.getrusage(RUSAGE_SELF).ru_maxrss` checkpoints
written **unbuffered to stderr** (buffered logging produced one wrong localisation
early on: a checkpoint that ran died with its line still in the buffer and looked
like it never executed).

**The kill follows a single large allocation, not a leak.** There is a slow creep of
~0.07 GB/min, which is not the cause:

```
08:57:35  rss_gb=4.36    <- creeping ~0.01 GB per 2 s
08:57:37  rss_gb=18.43   <- +14 GB in one 2 s interval
08:57:40  rss_gb=20.84   <- peak
08:57:44  process gone   <- exit 137
```

**Caveat on the peak, stated plainly:** observed peaks across runs ranged 10.9 to
24.9 GB, and in the lowest case free memory was already at 0.20 GB while the fit held
only 2.1 GB. Ambient memory pressure on this machine varies, so the absolute peak is
noisy and the kill threshold moves with it. What does not move is which
configurations fail.

**The allocation is in post-processing, after the sampler returns.** Checkpoints
around the first statements that touch `Posterior.samples`:

```
MEMCKPT sampler returned             peak_rss_gb=4.13
MEMCKPT got free_params              peak_rss_gb=4.13
MEMCKPT accessed .samples property   peak_rss_gb=4.13
MEMCKPT converted ONE value          peak_rss_gb=4.13
<killed>                             peak 10.88 GB
```

So accessing the property is cheap, and converting a single value is cheap. The
allocation happens while iterating the dict, even when the iteration is restricted to
the three sampled parameter names.

**The input is small by every cheap measure.** Reading `.shape` and `.nbytes` for all
entries is free and reports:

```
SAMPLES n_keys=40  total_mb=0.7
```

Forty keys for a model with **three** sampled parameters, 0.7 MB in total.

## Hypotheses

Not resolved. Two candidates, offered with the evidence rather than as conclusions:

1. **Per-draw materialisation that is not chunked.** The largest observed jump,
   ~14 GB, is about 2400 draws x 6 MB. A per-draw expansion whose size is set by the
   SSP grid would explain the exact grid-size dependence.
2. **Grid closure-captured rather than threaded.** `predict_state` and
   `predict_properties` expose `ssp_data=` / `template_data=` as a JIT-threading
   channel specifically because closure-captured grids misbehave under
   `jit`/`vmap`/`grad`. A post-fit path that wraps a model call without threading the
   grid would capture a copy per trace, and per-trace copies of a 109 MB grid is the
   right order of magnitude.

The forty-keys-for-three-parameters observation is worth chasing on its own: if the
extra entries are derived quantities that evaluate on access, then anything iterating
`samples` runs the forward model per draw, which needs the grid.

## What changed

This is a **regression**, not a platform limit. The same configuration on the same
109 MB C3K grid completed successfully on 2026-09-05 on tengri at the
`paper1/analysis` branch point (`6f52fb3f7`). It fails on `origin/main` (`5c476a03e`),
137 commits later.

Only two commits touch `src/tengri/inference/posterior.py` in that range:

- `3bc0506b1` fix(inference): dead-fit guards count every chain and skip Fixed
  parameters (#2087, #2088) - changes how `samples` is iterated and adds a
  free-parameter filter
- `f2b4b1843` AGNfitter-rX validation round (#2291)

`3bc0506b1` is the closer fit by area, though nothing in its diff obviously allocates.
A bisect over the 137 commits would settle it and is the cheapest next step.

## Workarounds tried, none sufficient

| attempt | result |
|---|---|
| `MAX_SAVED_DRAWS` 4000 -> 1000, `PPD_N_DRAWS` 200 -> 50 | no change; those constants are not on the failing path |
| `predict()` -> `predict_properties()` in the derived loop | no change; peak **bit-identical** (22.59 vs 22.58 GB) |
| bypass `Posterior.rhat()` with a local split-R-hat | 24.9 -> 20.7 GB, still killed |
| restrict the rhat input to `spec.free_params` | 20.8 GB, still killed |

The bit-identical peak in row two is worth noting: that is the signature of code that
never ran, and it is what showed the problem was upstream of where I was looking.

## Why it hid

Every cheap check passes. The fit converges, the diagnostics are good, `.nbytes`
reports 0.7 MB, and the two small-grid configurations work every time. Nothing
surfaces until a large grid meets a real posterior, and the failure then discards work
that had already succeeded.

## Suggested direction

Chunk any per-draw materialisation over `samples`; make derived entries explicit rather
than materialised on access, so a caller asking for sampled parameters never pays for
derived ones; and consider a guard that raises when a post-fit conversion is about to
expand by four orders of magnitude, so this class of failure produces an error message
instead of a silent kill.

**Comment by suchethac** [comment]
## Update: the feature that landed in the regression window

A peer session working on `profile_mass` pointed me at `mass_profile.py`, and the
archaeology is decisive.

```
git cat-file -e 6f52fb3f7:src/tengri/inference/mass_profile.py  -> DOES NOT EXIST
git cat-file -e 5c476a03e:src/tengri/inference/mass_profile.py  -> EXISTS
```

`6f52fb3f7` is the branch point where this same configuration, on this same 109 MB
C3K grid, completed successfully on 2026-09-05. `5c476a03e` is the pinned main where
it fails 5/5. **The entire amplitude-marginalization feature landed inside the 137
commits between them**, which is the regression window this report identified.

Three observations line up with it:

- `fitter.py:3873` calls `finalize_profile_mass(self, result, key=key)`, and its
  docstring states it is *"Called once from `Fitter.run()`, immediately after the
  backend runner returns"*. That is exactly where the memory checkpoints stop.
- It *"merges it into `posterior.samples`"*, which accounts for the observation this
  report could not otherwise explain: `samples` carries **40 keys for a model with 3
  sampled parameters**, 0.7 MB by `nbytes`, yet iterating it allocates on the order of
  10 GB.
- The census in #2349 reports profiling engaging on **8 of 9** constructible recipes,
  so nearly every configuration reaches this path.

**Explicitly not claimed:** that #2351 caused this. `e20fa08eb` is not in `5c476a03e`;
it landed afterwards, so the 48 -> 384 node change cannot be the cause here. It is
worth checking in the other direction though: if the reinsertion path already
allocates per-draw arrays that scale with the SSP grid, an 8x node increase would make
large-grid fits worse rather than better.

**The question this narrows to:** does the mass reinsertion materialise anything per
posterior draw whose size scales with the SSP grid? That is the only mechanism I can
construct that is simultaneously deterministic in grid size, invisible to `nbytes`,
and absent before the feature existed.

**Comment by suchethac** [comment]
## Mechanism confirmed, with the line and the arithmetic

Confirmed by the session that owns this area. `mass_profile.py:1175` on `origin/main`:

```python
A_ref_all, a_star_all, _, ell_ref_all = jax.vmap(stats_one)(samples_no_mass)
```

`stats_one` -> `_profile_stats` -> `_predict_full_vector`: a complete forward
prediction, `jax.vmap`ped over **every posterior draw at once**. Under `jit` + `vmap`
each intermediate inside that forward pass acquires a leading draw axis, so the
SSP-dependent intermediates become `(n_draws, n_age, n_wave)`.

That accounts for all three things this report could not otherwise explain: it is
deterministic in the SSP grid, it is invisible to `nbytes` (the allocation is
transient inside the jitted program and never lands in `posterior.samples`), and it is
absent before `finalize_profile_mass` existed.

### The arithmetic closes

Predicted peak = `n_draws x n_age x n_wave x 8 bytes`, at `n_draws = 2400`
(4 chains x 600 draws):

| cfg | SSP grid | n_age | n_wave | predicted | observed |
|---|---|---|---|---|---|
| V | `pgny_mist_c3k_chabrier` | 107 | 16902 | **32.3 GB** | 0/5 killed |
| I | `bc03_pdva_stelib_chabrier` | 221 | 6900 | **27.3 GB** | 0/5 killed |
| III | `fsps_mist_c3k_a_chabrier` | 107 | 11149 | **21.3 GB** | 0/5 killed |
| IV | `fsps_mist_c3k_a_chabrier` | 107 | 11149 | **21.3 GB** | 0/5 killed |
| II | `ssp_prsc_miles_chabrier_wNE` | 93 | 5994 | 10.0 GB | 5/5 passed |
| VI | `bpss_stars_c3k_a_chabrier` | 51 | 1221 | 1.1 GB | passed |

Sorted by predicted footprint, the table separates pass from fail with a clean
threshold between 10.0 and 21.3 GB, which is where a 48 GB machine with ~20 GB free
would be expected to cut. The measured peak for Config IV was 20.8-24.9 GB against a
prediction of 21.3 GB.

**This corrects a framing error in the original report.** I described the pattern as
depending on SSP *file* size. It does not: BC03/STELIB is a 35 MB file and fails,
while PARSEC/MILES is 64 MB and passes. The discriminator is `n_age x n_wave`; BC03
carries 221 age nodes. File size was a red herring.

### Not #2351

Re-stated because the original report raised it: `e20fa08eb` is not in `5c476a03e` and
cannot be the cause. The 48 -> 384 node change is on the *second* vmap
(`_sample_log_mass`), which is `O(nodes)` scalar work per draw: `(2400, 384)` floats,
about 7 MB. It is single-digit MB against a multi-GB problem, not a multiplier on it.

### A documented caveat with the wrong shape

`docs/dev/inference_methods.md:1666` does note a memory spike here, but describes it as
`(n_draws, n_pixels)` — the *output* — and scopes it to spectroscopy. The output is the
small part; the forward model's internal SSP-shaped intermediates are the cost, and
they apply to photometry too. 0/5 on three separate photometric configurations is the
measurement that shows the documented shape understates this.

### Fix direction

Chunking, not reduction. The algorithm genuinely needs one forward evaluation per
draw; it does not need all of them resident simultaneously. Bounded chunks
(`jax.lax.map` or an explicit loop over slices) cap peak at `chunk x intermediate`
rather than `n_draws x intermediate`, at no cost in correctness and little in speed,
since the per-chunk program stays compiled under the same engine key.
